Optimize bytecode root scanning

gasche opened the pull request:
This PR is a follow-up to #11169

@kayceesrk noticed a bytecode-interpreter performance regression in OCaml 5, with a large performance regression on the [Sandmarks stacks.ml benchmark](https://github.com/ocaml-bench/sandmark/blob/main/benchmarks/simple-tests/stacks.ml) (try with parameters `./bench 4000 floats-large`), where the bytecode-compiled version is 50% slower than the 4.13 one.

What is slow in trunk on this program is root scanning in the minor GC. The bytecode stack contains "naked" code pointers, and so during root scanning the runtime needs to check whether a pointer is a code pointer. This queries the code-fragments skiplist, which is slow for a very frequent operation.

#11169 is based on representing naked code pointers as immediates, but this makes backtraces imprecise so it did not gather consensus.

The present PR is based on the following remark from the 4.13 codebase:

https://github.com/ocaml/ocaml/blob/bfc5de586cfd3d8e28e99473c80c5e1c9d13de6f/runtime/roots_byt.c#L45-L51

In general, root scanning requires a code-fragment check, but in the specific case of `oldify_one` we don't need a check, because `oldify_one` itself checks for "being in the minor heap", which in particular implies "not being a code fragment".

By disabling the code-fragment check in this case, the present PR recovers exactly the 4.13 performance.

The current implementation is an ugly hack, mostly meant to generate discussion from more knowledgeable people: I optimize the code-fragment check away by checking whether the `scanning_action` pointer is exactly `&oldify_one`. Ugly! But it is not clear how to do better. We could make `scanning_action` a structure with metadata along the code pointer, the metadata being (for now) a flag "are you oldify_one?" or, more generally, "do you support naked pointers?".

gasche commented:
Note: in the past I assumed that the 4.13->trunk slowdown on this workload was caused by moving the code-fragment structures from a sequential skiplist to a lock-free skiplist, which adds synchronization overhead. But this guess is completely wrong: reverting the trunk implementation to an unsynchronized sequential skiplist keeps it just as slow. The problem comes from the fact that the root-scanning code has been refactored (call stacks can now be included in fibers so several different places need to scan them), and the new code structure lost the opportunity to easily special-case the oldification of the call stack to not check the code-fragment structures at all.

kayceesrk commented on runtime/fiber.c:
```diff
@@ -353,8 +353,9 @@ CAMLprim value caml_ensure_stack_capacity(value required_space)
-  return Is_block(v) && caml_find_code_fragment_by_pc((char *) v) == NULL;
+Caml_inline int is_block_and_not_code_frag(scanning_action f, value v) {
+  return Is_block(v) &&
+      (caml_is_oldify_one(f) || caml_find_code_fragment_by_pc((char *) v) == NULL);
```
There are repeated calls to `caml_is_oldify_one` in `caml_scan_stack`. In the ideal case, since this information doesn't change in a given call to `caml_scan_stack`, exactly one call is sufficient. Does the compiler already optimise it this way?

gasche replied:
@kayceesrk sorry for the naive question, but (assuming I mostly don't read assembly code) what's the equivalent of `-dcmm` for your favorite C compiler that I can use to answer this question?

gasche replied:
(With the new implementation using flags I'm very confident that the check will be optimized away, because it's not a function call to a separate object anymore. But it would still be nice to have a look.)

xavierleroy commented:
This is a nice idea, and a nice speed-up, but it deserves a better implementation than comparing the scanning action with a known function address.  (Obligatory reference: [Intentional equality for continuations](http://www.cs.princeton.edu/~appel/papers/conteq.pdf).)  We could go back to the 4.x approach, with a dedicated `caml_oldify_minor_roots` function, or add a `flags` argument to functions like `caml_scan_stack` to precise which kinds of roots are to be passed to the `scanning_action` parameter.

gasche committed:
[minor] refactor heap-verification code / API boundary

gasche committed:
optimize roots-scanning for the bytecode stack

The bytecode stack may contain naked code pointers. In general,
when scanning the stack to call scanning actions, we must check
the code-fragment table to filter out those naked pointers, but this
is fairly expensive.

In the specific case of the minor gc, we are scanning the stack with
[oldify_one], which gracefully accepts naked pointers (it ignores
all pointers outside the minor heap).

The current PR passes "scanning action flags" to all root scanners
(including user-provided hooks), to track whether their scanning
actions supports naked pointers, and disable the code-fragment check
in this case.

This fixes a 50% performance slowdown of the bytecode interpreter
compared to OCaml 4 on some benchmarks with large stacks full of
non-immediate values, such as the stacks.ml benchmark of
Sandmark (in 'floats-large' mode).

gasche commented:
@xavierleroy I updated the PR along the lines you suggested: there is now a new `enum stack_scanning_flags` parameter that is passed to all roots-scanning functions that may scan call stacks.

I hesitated a bit among the many different choices. I started with a more general `root_scanning_flags` that would be passed to all root-scanning functions, but (1) this is more invasive, (2) in particular it requires breaking the scanning-hooks API. The current approach is less general/extensible (if we want to add unrelated flags, we will have to update the scope), but simpler.

(Is it plausible that root scanning hooks would benefit from the information that they don't need to carefully rule out naked pointers in some cases? Would that benefit justify breaking the hook-registration to add a new parameter? If someone has an opinion on this, it's probably @gadmm, so here is a cc:)

gasche commented:
Actually, it looks like I will need to extend root-scanning hooks with scanning flags anyway, given that the thread runtime registers its own scanning function as a hook. Otherwise we recover the very-slow behavior when scanning the stack of all inactive threads.

gasche commented:
I updated the PR with a general `enum scanning_action_flags` passed to all scanning actions, including users-provided scanning hooks.

gasche committed:
Changes

gadmm commented:
I am not sure what semantics you are trying to achieve with the flag about naked pointers. It works as it is, but it looks like grabbing the problem by the wrong end. All scanning functions accept naked pointers, either in the old sense for the "naked pointers" mode or in the restricted sense in "no-naked-pointers" mode. Your flag only makes sense in the context of a mismatch between the VM's naked pointers and the latter which is the source of the problem.

Instead of introducing a new kind of naked pointers in no-naked-pointers mode, the useful distinction is whether the scanning function is noop on non-young values or not. Armed with this knowledge, you can implement a more efficient "page table test" using `Is_young` before passing to the scanning function. We already use this distinction for optimisation in https://gitlab.com/ocaml-rust/ocaml-boxroot/-/blob/f15bc/boxroot/boxroot.c#L602-606 and we could very much make use of a predicate "this scanning function is a noop on non-young values" (whether in the form of a flag or not) instead of our current hack.

This is a useful information for users of scanning hooks, because the `Is_young` test is easier to optimise on the caller's side (where one can use tricks _à la_ Dolan). With the optimised test at https://gitlab.com/ocaml-rust/ocaml-boxroot/-/blob/f15bc/boxroot/boxroot.c#L588, we observed a 4-fold speedup when scanning pools that contains very few young values (and nice speedups in other cases). It might be worth looking at the existing callers of scanning functions to see whether opportunities in the runtime and the Coq VM exists for such optimisations. (If you further make it _mandatory_ for the caller to filter young blocks, then you can also remove the test inside `oldify_one`, but I do not know if it is interestingly more efficient to do so.)

Anyway it is a good time for such changes because all users of scanning hooks have to change their code for 5.0.

gasche commented:
> All scanning functions accept naked pointers, either in the old sense for the "naked pointers" mode or in the restricted sense in "no-naked-pointers" mode.

My understanding is that if you call `caml_darken` on a code pointer, you get a crash. In this sense, this function does not accept naked pointers (unless they have the shape of an already-black block, etc.).

My take-away from your feedback is that I should track `SCANNING_ONLY_YOUNG_VALUES` rather than `SCAN_ACCEPTS_NAKED_POINTERS`. I'm fine with this suggested change.

gasche commented:
I updated the code as suggested, and marked the PR "ready for review". (Hint hint.)

Note: I marked this as a 5.0 change, as it fixes a noticeable bytecode performance regression and it breaks the scanning-hooks API (already broken in 5.0).

kayceesrk commented on runtime/major_gc.c:
```diff
@@ -742,6 +742,8 @@ static intnat mark(intnat budget) {
   return budget;
 }
 
+static enum scanning_action_flags darken_scanning_flags = 0;
```
IIUC the state here is `!SCANNING_ONLY_YOUNG_VALUES`. For readability, would it not be better to have another constant in the `scanning_action_flags` enum?

gasche replied:
I prefer 0. I think of these flags as a set of atomic flags of value 2^k -- not an enum 1,2,3..., but really flags 1,2,4,8... By construction, 0 is the empty set; using it feels appropriate in connection with the test code (`flags & FOO`) which exposes the assumption that intersection is bitwise-and, and non-emptyness is non-0ness. Does that make sense?

kayceesrk replied:
Thanks for the clarification.

gadmm replied:
A bitfield is a good idea, but usually you still define 0 as "None" (or something) in the enum for this purpose.

kayceesrk commented on runtime/shared_heap.c:
```diff
@@ -728,6 +728,8 @@ void caml_verify_root(void* state, value v, value* p)
   verify_push(state, v, p);
 }
 
+static enum scanning_action_flags verify_scanning_flags = 0;
```
Same here.

gadmm commented on Changes:
```diff
@@ -124,6 +125,12 @@ OCaml 5.0
   `caml_process_pending*` for multicore.
   (Guillaume Munch-Maccagnoni, review by Sadiq Jaffer and Gabriel Scherer)
 
+* #11337: pass 'flags' metadata to root scanners, to optimize stack
```
The root-scanning hooks are internal, so this is not considered a breaking change.

gasche replied:
done, thanks!

gadmm commented on runtime/caml/roots.h:
```diff
@@ -21,16 +21,25 @@
 
+enum scanning_action_flags {
+  SCANNING_ONLY_YOUNG_VALUES = 1, // action is a no-op outside the minor heap
+};
```
```suggestion
typedef enum {
  SCANNING_GENERIC = 0, // no special property
  SCANNING_ONLY_YOUNG_VALUES = 1, // action is a no-op outside the minor heap
} scanning_action_flags;
```
How about a typedef to omit `enum` in signatures? (like other typedef enums in the runtime)

gasche replied:
I went for your `typedef` suggestion. (It's a question of style and I have a mild preference for what I had written, but it's good to put more weight in other people's expressed preference, and indeed it seems that this style dominates in existing code.)

gadmm commented on runtime/major_gc.c:
```diff
@@ -757,7 +759,8 @@ void caml_darken_cont(value cont)
         value stk = Field(cont, 0);
         if (Ptr_val(stk) != NULL)
-          caml_scan_stack(&caml_darken, 0, Ptr_val(stk), 0);
+          caml_scan_stack(&caml_darken, darken_scanning_flags, 0,
```
```suggestion
          caml_scan_stack(&caml_darken, SCANNING_GENERIC, 0,
```
etc. in other places, in accordance with the suggestion of having a zero field.

gasche replied:
I thought again about having a zero field, but I don't like it. (For once, I don't know of a good name for the field; `SCANNING_FLAGS_NONE` is the least meh but still not exciting.)

I went to look at how other flags-enum do this in the codebase, and it looks like some have an explicit name for the empty set, others don't. (For example the `CHANNEL` flags in caml/io.h don't have one, or the marshalling flags in `extern.c`, but the `SELECT_MODE` flags in unix/select_win32.c do have one.)

xavierleroy replied:
If your flags are a set (of logically-ORed powers of two), then `0` is perfect because it clearly denotes the empty set.  (Squint a bit and you'll see ∅.  Or remember your set theory class, the part where natural numbers are constructed.)

gadmm commented on runtime/minor_gc.c:
```diff
@@ -222,6 +222,10 @@ static int try_update_object_header(value v, value *p, value result,
 
+/* oldify_one is a no-op outside the minor heap. */
+static enum scanning_action_flags oldify_scanning_flags =
+  SCANNING_ONLY_YOUNG_VALUES;
```
I am not sure to see the interest of having a definition next to the function, it is clear enough already (as explained in the comment above).

gasche replied:
I like these separate definitions, it's a good place to have a comment about the choice, and it avoids duplicating the constant among several calls with the same scanning action. I propose to keep this part as-is in the "set of flags" option.

gadmm reviewed (approved):
This looks good to me. As explained I think that this is useful beyond the VM usage. Some minor comments below; I essentially feel that the definition of the flag next to the function is half-way between having a less verbose generic flag, and packing the function along with the flag (e.g. by turning `scanning_action` into a struct), and that either would be a better option.

gasche commented:
Could people with opinions converge on a consensus between passing the flags separately to each function, or building a struct that combines the function pointer with metadata? (In this case I don't need a set-of-flags value, each flag can be a boolean member of the struct.) I'm happy to implement either, but it's some work so I would like to know what people prefer.

gadmm commented:
Both are fine to me, it is as you prefer (and at this stage "it's less work for me" is fine as a reason). (In general if I check "accept" in my review it means that the remaining suggestions are up to the author.)

gasche commented:
I made changes according to Guillaume's feedback, still on the "separate flags" route.

I don't mind trying to implement the "struct" route instead (this was my initial idea for an API, but Xavier rather mentioned a separate flags parameter so I went for that, it sounded more low-tech), if people have a preference for the struct. Any thoughts?

sadiqj reviewed (approved):
Little late but this looks good to me.

sadiqj commented:
> Any thoughts?

Feels a tiny bit weird that the flag is a property of the action and yet separate but I'm not sure it's worth a change to a struct.

ctk21 commented:
Taken a look, I believe it implements what's intended and should avoid the code frag check when passing `oldify_one` for bytecode (there is no change for native). Neat speedup for bytecode.

gasche commented:
I started working on a `struct` solution but I was struck by decision fatigue ; uncertainty about C API design. I have to decide whether the functions would receive the struct (packing the scanning action code pointer and a boolean field) by value or by pointer:
- by value: this is never done in the runtime so far, I vaguely remember hearing about ABI stability/portability issues with structs-by-value, and I don't know if adding more fields later is going to cause a performance issue
- by pointer: unless I'm mistaken, this adds an extra indirection in a fairly hot path of the runtime, and I don't know how to measure the performance impact

Unless people have clear ideas on which way to go, I vote for sticking with the current separate-flags-parameter route for now. (As Guillaume kindly pointed out, those hooks are external anyway so we could consider changing later again if we want.)

gasche commented:
@Octachron I am planning to cherry-pick this in 5.0 (it's fixing a performance regression of OCaml 5). Does that work for you?

gadmm commented:
I think the current solution is good enough and everyone seems to agree to move towards a merge.

FWIW I do not think there are any performance concerns, the hooks are not called often. I do not think there is any ABI stability guarantee in the runtime, but then there may anyway be a policy against struct by value.

gasche commented:
> FWIW I do not think there are any performance concerns, the hooks are not called often.

The user-defined hooks are not called often but here we are talking about an extra indirection in the call to `oldify_one` by the local-roots scanner, and that sounds performance-sensitive to me.

gasche commented:
> I think the current solution is good enough and everyone seems to agree to move towards a merge.

This is also my understanding, thanks!

gasche commented:
Cherry-picked in 5.0 as a87f4a023034adf6949d3a8410d9a33aa9e0ce9a.
